aiaccel/hpo/samplers/nelder_mead_sampler.py

```diff
@@ -0,0 +1,191 @@
+from __future__ import annotations
+
+import queue
+import warnings
+from collections.abc import Generator
+from dataclasses import dataclass
+from typing import Any, Sequence
+
+import numpy as np
+import optuna
+from optuna.distributions import BaseDistribution
+from optuna.study import Study
+from optuna.trial import FrozenTrial, TrialState
+
+
+@dataclass
+class NelderMeadCoefficient:
+    r: float = 1.0
+    ic: float = -0.5
+    oc: float = 0.5
+    e: float = 2.0
+    s: float = 0.5
+
+
+class NelderMeadAlgorism:
+    vertices: np.ndarray[Any, Any]
+    values: np.ndarray[Any, Any]
+
+    def __init__(
+        self,
+        search_space: dict[str, list[float]],
+        coeff: NelderMeadCoefficient | None = None,
+        seed: int | None = None,
+    ) -> None:
+        self._search_space = search_space
+        self.coeff = coeff if coeff is not None else NelderMeadCoefficient()
+
+        self.dimension = len(search_space)
+
+        self.value_queue: queue.Queue[float] = queue.Queue()
+        self._rng: np.random.RandomState = np.random.RandomState(seed)
+        self.is_ready: bool = False
+
+    def __iter__(self) -> Generator[np.ndarray[Any, Any], None, None]:
+        # initialization
+        lows, highs = zip(*self._search_space.values())
+        self.vertices = self._rng.uniform(lows, highs, (self.dimension + 1, self.dimension))
+
+        self.is_ready = True
+        yield from iter(self.vertices[:-1])
+        self.is_ready = False
+        yield self.vertices[-1]
```

Comment:
後ろにもこの形が出てきて結構複雑になってしまいますね…
```suggestion
        yield from self.yield_vertices(self.vertices)
```
みたいなものを作って，全てのyield, yield fromをここに詰める，引数に含まれる要素数に合わせてself.is_readyを適宜制御する実装はいかがでしょうか？

Reply:
NelderMeadAlgorism.yield_vertices を追加しました。

d4097e23c9ada3b49115a19b3cbeba2fc9d854c8